runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamFlinkDataStreamAdapter.java

```diff
@@ -0,0 +1,302 @@
+                    Collector<WindowedValue<InputT>> out)
+                    throws Exception {
+                  out.collect(
+                      WindowedValue.timestampedValueInGlobalWindow(
+                          value,
+                          ctx.timestamp() == null
+                              ? BoundedWindow.TIMESTAMP_MIN_VALUE
+                              : Instant.ofEpochMilli(ctx.timestamp())));
+                }
+              },
+              BeamAdapterCoderUtils.coderToTypeInformation(
+                  WindowedValue.getFullCoder(
+                      BeamAdapterCoderUtils.typeInformationToCoder(
+                          flinkInput.getType(), coderRegistry),
+                      GlobalWindow.Coder.INSTANCE),
+                  pipelineOptions)));
+    };
+  }
+
+  private <InputT>
+      FlinkStreamingPortablePipelineTranslator.PTransformTranslator<
+              FlinkStreamingPortablePipelineTranslator.StreamingTranslationContext>
+          flinkOutputTranslator(Map<String, DataStream<?>> outputMap) {
+    return (String id,
+        RunnerApi.Pipeline p,
+        FlinkStreamingPortablePipelineTranslator.StreamingTranslationContext context) -> {
+      // When we run into a FlinkOutput operator, we cache the computed PCollection to return to the
+      // user.
+      RunnerApi.PTransform transform = p.getComponents().getTransformsOrThrow(id);
+      DataStream<WindowedValue<InputT>> inputDataStream =
+          context.getDataStreamOrThrow(Iterables.getOnlyElement(transform.getInputsMap().values()));
+      String outputId = transform.getSpec().getPayload().toStringUtf8();
+      Coder<InputT> outputCoder =
+          BeamAdapterCoderUtils.lookupCoder(
+              p, Iterables.getOnlyElement(transform.getInputsMap().values()));
+      // TODO(robertwb): Also handle or disable length prefix coding (for embedded mode at least).
+      outputMap.put(
+          outputId,
+          inputDataStream.transform(
+              "StripWindows",
+              BeamAdapterCoderUtils.coderToTypeInformation(outputCoder, pipelineOptions),
+              new UnwrapWindowOperator<InputT>()));
+    };
+  }
+
+  /**
+   * Forwards the Beam timestamps to the underlying Flink timestamps, but unlike {@link
+   * DataStream#assignTimestampsAndWatermarks(WatermarkStrategy)} does not discard the underlying
+   * watermark signals.
+   *
+   * @param <T> user element type
+   */
+  private static class UnwrapWindowOperator<T> extends AbstractStreamOperator<T>
+      implements OneInputStreamOperator<WindowedValue<T>, T> {
+    @Override
+    public void processElement(StreamRecord<WindowedValue<T>> element) {
+      output.collect(
+          element.replace(
+              element.getValue().getValue(), element.getValue().getTimestamp().getMillis()));
+    }
+  }
+
+  private <K, V> V getNonNull(Map<K, V> map, K key) {
+    return Preconditions.checkNotNull(map.get(Preconditions.checkNotNull(key)));
```

Comment:
You'll want to use Beam's `checkArgumentNotNull` or `checkStateNotNull` with an explicit error message to indicate the source of the problem. This one from Guava will throw an NPE again bringing users knocking on our door about it being an unreasonable crash with no debugging info.

Reply:
Done.